src/GafferOSL/OSLCode.cpp

```diff
@@ -0,0 +1,416 @@
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include <fstream>
+
+#include "boost/filesystem.hpp"
+#include "boost/bind.hpp"
+
+#include "OSL/oslcomp.h"
+
+#include "IECore/Exception.h"
+
+#include "Gaffer/StringAlgo.h"
+#include "Gaffer/StringPlug.h"
+#include "Gaffer/Metadata.h"
+#include "Gaffer/SplinePlug.h"
+#include "Gaffer/Process.h"
+
+#include "GafferOSL/Private/CapturingErrorHandler.h"
+#include "GafferOSL/OSLCode.h"
+
+using namespace std;
+using namespace IECore;
+using namespace OSL;
+using namespace Gaffer;
+using namespace GafferOSL;
+
+//////////////////////////////////////////////////////////////////////////
+// Code generation
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+string colorSplineParameter( const SplinefColor3fPlug *plug )
+{
+	string result;
+	result += "\tfloat " + plug->getName().string() + "Positions[] = { 0, 0, 1, 1 },\n";
+	result += "\tcolor " + plug->getName().string() + "Values[] = { 0, 0, 1, 1 },\n";
+	result += "\tstring " + plug->getName().string() + "Basis = \"catmull-rom\",\n";
+	return result;
+}
+
+string parameter( const Plug *plug )
+{
+	const Gaffer::TypeId plugType = (Gaffer::TypeId)plug->typeId();
+	if( plugType == SplinefColor3fPlugTypeId )
+	{
+		return colorSplineParameter( static_cast<const SplinefColor3fPlug *>( plug ) );
+	}
+
+	string type;
+	string defaultValue;
+	switch( plugType )
+	{
+		case FloatPlugTypeId :
+			defaultValue = "0.0";
+			type = "float";
+			break;
+		case IntPlugTypeId :
+			defaultValue = "0";
+			type = "int";
+			break;
+		case Color3fPlugTypeId :
+			defaultValue = "color( 0.0 )";
+			type = "color";
+			break;
+		case V3fPlugTypeId :
+			defaultValue = "vector( 0.0 )";
+			type = "vector";
+			break;
+		case M44fPlugTypeId :
+			defaultValue = "1";
+			type = "matrix";
+			break;
+		case StringPlugTypeId :
+			defaultValue = "\"\"";
+			type = "string";
+			break;
+		case PlugTypeId :
+			defaultValue = "0";
+			type = "closure color";
+			break;
+		default :
+			throw IECore::Exception( string( "Unsupported plug type \"" ) + plug->typeName() + "\"" );
+	}
+
+	string direction = plug->direction() == Plug::Out ? "output " : "";
+	return "\t" + direction + type + " " + plug->getName().string() + " = " + defaultValue + ",\n";
+}
+
+string generate( const OSLCode *shader, string &shaderName )
+{
+	// Start with parameters
+
+	string result;
+
+	for( PlugIterator it( shader->parametersPlug() ); !it.done(); ++it )
+	{
+		result += parameter( it->get() );
+	}
+
+	result += "\n";
+
+	for( PlugIterator it( shader->outPlug() ); !it.done(); ++it )
+	{
+		result += parameter( it->get() );
+	}
+
+	result += ")\n";
+
+	// Add on body
+
+	const std::string code = shader->codePlug()->getValue();
+	result += "{\n" + code;
```

Comment:
Me too, but I was hoping to avoid having to get into too much fiddling with the text in case there are fragile corner cases I don't know about (multiline strings maybe?).